crates/core/src/routing_info/mod.rs

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) 2024 -  Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::HashMap;
+use std::pin::pin;
+use std::sync::Arc;
+
+use arc_swap::ArcSwap;
+use tokio::sync::mpsc;
+use tokio::time::MissedTickBehavior;
+use tracing::{debug, info};
+use xxhash_rust::xxh3::Xxh3Builder;
+
+use restate_types::cluster_controller::SchedulingPlan;
+use restate_types::config::Configuration;
+use restate_types::identifiers::PartitionId;
+use restate_types::metadata_store::keys::SCHEDULING_PLAN_KEY;
+use restate_types::{NodeId, Version, Versioned};
+
+use crate::metadata_store::MetadataStoreClient;
+use crate::{
+    cancellation_watcher, task_center, ShutdownError, TaskCenter, TaskHandle, TaskId, TaskKind,
+};
+
+pub type CommandSender = mpsc::Sender<Command>;
+pub type CommandReceiver = mpsc::Receiver<Command>;
+
+pub enum Command {
+    /// Request an on-demand refresh of routing information from the authoritative source.
+    SyncRoutingInformation,
+}
+
+/// Holds a view of the known partition-to-node mappings. Use it to discover which node(s) to route
+/// requests to for a given partition. Compared to the partition table, this view is more dynamic as
+/// it changes based on cluster nodes' operational status. This handle can be cheaply cloned.
+#[derive(Clone)]
+pub struct PartitionRouting {
+    sender: CommandSender,
+    /// A mapping of partition IDs to node IDs that are believed to be authoritative for that serving requests.
+    partition_to_node_mappings: Arc<ArcSwap<PartitionToNodesRoutingTable>>,
+}
+
+impl PartitionRouting {
+    /// Look up a suitable node to answer requests for the given partition.
+    pub fn get_node_by_partition(&self, partition_id: PartitionId) -> Option<NodeId> {
+        self.partition_to_node_mappings
+            .load()
+            .inner
+            .get(&partition_id)
+            .copied()
+    }
+
+    /// Call this to hint to the background refresher that its view may be outdated. This is useful
+    /// when a caller discovers via some other mechanism that routing infromation may be invalid -
+    /// for example, when a request to a node previously returned by `get_node_by_partition` fails
+    /// with a response that explicitly indicates that it is no longer serving that partition. The
+    /// call returns as soon as the request is enqueued. A refresh is not guaranteed to happen.
+    pub async fn request_refresh(&self) {
+        self.sender
+            .send(Command::SyncRoutingInformation)
+            .await
+            .expect("Failed to send refresh request");
```

Comment:
Nit: Since the `Command` no longer carries any specific value, one could use a watch to signal the refresh request. That way, the `request_refresh` call could become synchronous and users of this method wouldn't block on the availability of send permits.

Reply:
That's a neat use of watch! I do think the sync command will grow to carry some information again in the future - perhaps a node id, partition id, or some other context from the caller - as we learn more about the conditions that should trigger a refresh. So, I'd rather keep in as is for the time being, even though it's a totally valid nit!